test that sibling $ids to $refs are ignored

karenetheridge opened the pull request:
This tests that $ref resolves to something different in draft3-7 than one would expect in modern versions, because the $id is not parsed when it is adjacent to a $ref. This is *not* the same as a $ref not being resolvable at all.

jdesrosiers commented:
This is more controversial that you might think. See #458. In the end we decided to consider this undefined behavior and it should not have tests. #492 is good because it changes the test to avoid the ambiguous case. This, however, is not going to go over well with some people and is contrary to the precedent set in #458.

karenetheridge commented:
I agree with the changes in #458 - that a $ref into a definitions adjacent to a $ref should not be resolveable as that part of the schema "doesn't exist". I even have a test for this in my draft7 implementation that fails (and marked TODO, but in reality I never intend to fix this) because of the extra complexity involved - either we mutate the schema and physically remove the unaccessable bits, or we keep track of these unaccessable locations in a separate list and check it whenever navigating a $ref to see if that path falls in the verboten area.

However, I don't think the changes in this PR (#493) are the same thing at all. It's much more straightforward when processing any keyword (or when starting out processing a subschema, and considering what keywords to look at) to go "there's a $ref here, so we won't process any other keywords adjacent to it".

What am I missing? what about the changes here in #493 are controversial/ambiguous?

jdesrosiers commented:
I think this change represents how it should be interpreted. @Julian was the one who disagreed with my interpretation in #458, so maybe he will have something to say about this.

> I don't think the changes in this PR (#493) are the same thing at all.

I don't know what to say to this. It's exactly the same case as #458. The only difference you've identified is that some keywords are easier to ignore in your implementation that others. In my implementation, ignoring keywords inside a reference is trivial. It's just a natural consequence of the architecture.

If anything, `$id` and `$schema` are the ones that it makes sense to make an exception for. Otherwise it's extremely awkward to figure out how to interpret a reference. A dialect needs to be declared somehow in order to know what version of `$ref` it is. Without allowing a `$schema`, the only way to know is if the implementation allows you to specify a default dialect. `$id` is similar. The only way to set a base URI is if the implementation allows you to set it externally. Without a base URI, any relative non-fragment-only URI would be invalid and cause an error.

karenetheridge commented:
> It's exactly the same case as #458

No, it's not, because it's a different keyword, that's evaluated differently.  That's exactly why we have different tests for different combinations of keywords, because we don't know how implementation decisions might affect how different combinations of keywords interact.  Some combinations seem similar enough that we don't bother testing all variants, but as various implementations pop up where it's apparent that the differences are relevant, we add tests for them. Dismissing them because they're "exactly the same" to you and the architecture decisions you made is not constructive.

Evaluating an $id (or not) when it is adjacent to a $ref can have a very different implementation compared to processing a $ref from a totally different location to a position that is not adjacent to but rather a nephew to another $ref.  The target subschema is not the current subject of the evaluation - not the uncle "$ref", not even the parent "definition" that contains the resolved location.

I can think of a *third* implementation architecture (different from both mine and ours) that might get tripped up by the test I added here in this PR, because of the way it pre-scans a schema for $ids before it starts the recursive evaluation phase, where it might not properly detect that there is a sibling $ref and therefore know to ignore the keyword. Another case that would be similarly caught is covered by another test case I added a while ago in #484 -- which is actually covered by a totally different part of the spec but because of particular architecture decisions might fail due to the same code implementation.

jdesrosiers commented:
> It's exactly the same case as #458
>
> > No, it's not, because it's a different keyword, that's evaluated differently.

In #458 we didn't agree on an interpretation. We compromised by agreeing to consider the case undefined and should not be tested either way. If the behavior of `definitions` in a `$ref` is considered undefined, why would it be any different for `$id`? Again, you and I agree on the correct way to handle this. I only bring this up to be respectful of others who have expressed disagreement on this interpretation in the past and to stick to the agreement we made. But maybe I'm misunderstanding the opposing interpretation. I'll shut up now and let others speak for themselves.

> Dismissing them because they're "exactly the same" to you and the architecture decisions you made is not constructive.

This is a misunderstanding. You seemed to be making an argument based on properties of your implementation. I mentioned how the properties of my implementation differ in order to show that any argument based on the properties of any implementation is not helpful. I was quite surprised to see the same argument directed back at me.

karenetheridge commented:
Okay. If we're all satisfied that the existing tests are okay the way they are, we don't need to belabour this point.

Julian commented:
To me yes this is likely equally undefined, but as I said somewhere, I'm likely tired of arguing about this topic :).
